superqt/signal/_signal.py

```diff
@@ -0,0 +1,181 @@
+from typing import TYPE_CHECKING, Any, Callable, Union, overload
+
+if TYPE_CHECKING:
+    CallbackType = Callable[..., None]
+    SlotRef = Union[CallbackType, weakref.WeakKeyDictionary[object, CallbackType]]
+
+
+class Signal:
+
+    # valid callback signatures for this signal.
+    if TYPE_CHECKING:
+        signatures: tuple[Signature, ...]
+        _signal_instances: dict[Signal, weakref.WeakKeyDictionary[Any, SignalInstance]]
+
+    def __init__(self, *types: Any) -> None:
+        self._signal_instances = {}
+
+        if not types:
+            # single signature, no parameters
+            self.signatures = (Signature(),)
+        elif not isinstance(types[0], (list, tuple, Signature)):
+            # single signature, with parameters
+            self.signatures = (Signal._build_signature(types),)
+        else:
+            # multiple signatures
+            _signatures: list[Signature] = []
+            for t in types:
+                if isinstance(t, Signature):
+                    _signatures.append(t)
+                elif isinstance(t, (list, tuple)):
+                    _signatures.append(Signal._build_signature(t))
+                else:
+                    raise TypeError(
+                        "each signature in a multi-signature sequence must be "
+                        "experessed as a Signature, list, or tuple"
+                    )
+            self.signatures = tuple(_signatures)
+
+    @staticmethod
+    def _build_signature(types: Sequence[type]) -> Signature:
+        return Signature(
+            [
+                Parameter(name=f"a{i}", kind=Parameter.POSITIONAL_ONLY, annotation=t)
+                for i, t in enumerate(types)
+            ]
+        )
+
+    def __getattr__(self, name):
+        if name == "connect":
+            name = self.__class__.__name__
+            raise AttributeError(
+                f"{name!r} object has no attribute 'connect'. You can connect to the "
+                "signal on the *instance* of a class with a Signal() class attribute. "
+                "Or create a signal instance directly with SignalInstance."
+            )
+        return self.__getattribute__(name)
+
+    @overload
+    def __get__(self, instance: None, owner: type | None = None) -> Signal:
+        ...
+
+    @overload
+    def __get__(self, instance: Any, owner: type | None = None) -> SignalInstance:
+        ...
+
+    def __get__(self, instance: Any, owner: type = None) -> Signal | SignalInstance:
+        # if instance is not None, we're being accessed on an instance of `owner`
+        # otherwise we're being accessed on the `owner` itself
+        if instance is None:
+            return self
+        d = self._signal_instances.setdefault(self, weakref.WeakKeyDictionary())
+        return d.setdefault(instance, SignalInstance(self.signatures, instance))
+
+
+class SignalInstance:
+    def __init__(
+        self, signatures: tuple[Signature, ...] = (), instance: Any = None
+    ) -> None:
```

Comment:
is the difference between `Signal` and `SignalInstance` that `Signal` is the type of a signal and the other is a specific instance of a given signal?

Reply:
yeah, pretty much.  it's confusing... but interesting! and get's to the heart of python's [descriptor protocol](https://docs.python.org/3/howto/descriptor.html) (the same thing behind the `@property` decorator).  but, consider this code with standard Qt signals:

```py
In [1]: from qtpy.QtCore import QObject, Signal

In [2]: class MyObj(QObject):
   ...:     changed = Signal() 
   ...:

In [3]: my_obj = MyObj()

In [4]: type(MyObj.changed)
Out[4]: <class 'PySide2.QtCore.Signal'>

In [5]: type(my_obj.changed)
Out[5]: <class 'PySide2.QtCore.SignalInstance'>
```

`changed` is a class attribute (i.e. it doesn't know about self).  When you access the `changed` attribute on the `MyObj` class, you get an instance of the `Signal` type... like you'd expect. 

`Signal` however, implements the descriptor protocol (it provides a [`__get__` method](https://docs.python.org/3/reference/datamodel.html#object.__get__)), which lets you detect whether the attribute is being accessed on the Class itself, or an instance of the class (`my_obj`).  If we are being accessed on an instance, we return a new object that handles all of the callback connections to _that specific_ instance.

This is what allows different instances of MyObj to have their own callbacks connected:

```py
# two instances
my_obj = MyObj()
my_other_obj = MyObj()

# one connected
my_obj.changed.connect(do_something)

# other emits
my_other_obj.changed.emit()

# nothing happens... because the my_other_obj.changed SignalInstance has no callbacks
```